File: src/main/java/net/ndrei/teslacorelib/inventory/MultiItemHandler.java
package net.ndrei.teslacorelib.inventory;

import com.google.common.collect.Lists;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.IItemHandlerModifiable;
import net.ndrei.teslacorelib.compatibility.ItemStackUtil;

import javax.annotation.Nonnull;
import java.util.List;

/**
 * Created by CF on 2016-12-16.
 */
public class MultiItemHandler implements IMultiItemHandler {
    private List<IItemHandler> handlers = null;

    public MultiItemHandler(List<IItemHandler> handlers) {
        if (null == (this.handlers = handlers)) {
            this.handlers = Lists.newArrayList();
        }
    }

    public void addItemHandler(IItemHandler handler) {
        this.handlers.add(handler);
    }

    @Override
    public boolean canInsertItem(int slot, ItemStack stack) {
        if (slot < 0) {
            return false;
        }
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            if (handler.getSlots() > slot) {
                if (handler instanceof IFilteredItemHandler) {
                    // filter the inputs
                    return ((IFilteredItemHandler)handler).canInsertItem(slot, stack);
                }
                return true;
            }
            slot -= handler.getSlots();
        }
        return false;
    }

    @Override
    public boolean canExtractItem(int slot) {
        if (slot < 0) {
            return false;
        }
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            if (handler.getSlots() > slot) {
                if (handler instanceof IFilteredItemHandler) {
                    // filter the inputs
                    return ((IFilteredItemHandler)handler).canExtractItem(slot);
                }
                return true;
            }
            slot -= handler.getSlots();
        }
        return false;

    }

    @Override
    public int getInventories() {
        return this.handlers.size();
    }

    @Override
    public IItemHandler getInventory(int inventory) {
        return this.handlers.get(inventory);
    }

    @Override
    public IFilteredItemHandler getFilteredInventory(int inventory) {
        IItemHandler handler = this.getInventory(inventory);
        if (handler instanceof IFilteredItemHandler) {
            return (IFilteredItemHandler) handler;
        }
        return null;
    }

    @Override
    public boolean canInsertItem(int inventory, int slot, ItemStack stack) {
        IFilteredItemHandler handler = this.getFilteredInventory(inventory);
        if (handler != null) {
            return handler.canInsertItem(slot, stack);
        }
        return true;
    }

    @Override
    public boolean canExtractItem(int inventory, int slot) {
        IFilteredItemHandler handler = this.getFilteredInventory(inventory);
        if (handler != null) {
            return handler.canExtractItem(slot);
        }
        return true;
    }

    @Override
    public int getSlots() {
        int slots = 0;
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            slots += handler.getSlots();
        }
        return slots;
    }

    @Nonnull
    @Override
    public ItemStack getStackInSlot(int slot) {
        if (slot < 0) {
            return ItemStackUtil.getEmptyStack();
        }
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            if (handler.getSlots() > slot) {
                return handler.getStackInSlot(slot);
            }
            slot -= handler.getSlots();
        }
        return ItemStackUtil.getEmptyStack();
    }

    @Nonnull
    @Override
    public ItemStack insertItem(int slot, @Nonnull ItemStack stack, boolean simulate) {
        if (slot < 0) {
            return ItemStackUtil.getEmptyStack();
        }
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            if (handler.getSlots() > slot) {
                return handler.insertItem(slot, stack, simulate);
            }
            slot -= handler.getSlots();
        }
        return ItemStackUtil.getEmptyStack();
    }

    @Nonnull
    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        if (slot < 0) {
            return ItemStackUtil.getEmptyStack();
        }
        for(int i = 0; i < this.handlers.size(); i++) {
            IItemHandler handler = this.handlers.get(i);
            if (handler.getSlots() > slot) {
                return handler.extractItem(slot, amount, simulate);
            }
            slot -= handler.getSlots();
        }
        return ItemStackUtil.getEmptyStack();
    }

//    @Override
//    public int getSlotLimit(int slot) {
//        if (slot < 0) {
//            return 0;
//        }
//        for(int i = 0; i < this.handlers.size(); i++) {
//            IItemHandler handler = this.handlers.get(i);
//            if (handler.getSlots() > slot) {
//                return handler.getSlotLimit(slot);
//            }
//            slot -= handler.getSlots();
//        }
//        return 0;
//    }

    @Override
    public void setStackInSlot(int slot, @Nonnull ItemStack stack) {
        if (slot >= 0) {
            for (int i = 0; i < this.handlers.size(); i++) {
                IItemHandler handler = this.handlers.get(i);
                if (handler.getSlots() > slot) {
                    if (handler instanceof IItemHandlerModifiable) {
                        ((IItemHandlerModifiable)handler).setStackInSlot(slot, stack);
                    }
                    else {
                        throw new RuntimeException("Target inventory is not an IItemHandlerModifiable.");
                    }
                    return;
                }
                slot -= handler.getSlots();
            }
        }
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/inventory/MultiItemHandler.kt
package net.ndrei.teslacorelib.inventory

import net.minecraft.item.ItemStack
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.IItemHandlerModifiable

/**
 * Created by CF on 2017-06-28.
 */
open class MultiItemHandler(private val handlers: MutableList<IItemHandler> = mutableListOf())
    : IMultiItemHandler {

    fun addItemHandler(handler: IItemHandler) {
        this.handlers.add(handler)
    }

    override fun canInsertItem(slot: Int, stack: ItemStack): Boolean {
        var currentSlot = slot
        if (currentSlot < 0) {
            return false
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                if (handler is IFilteredItemHandler) {
                    // filter the inputs
                    return handler.canInsertItem(currentSlot, stack)
                }
                return true
            }
            currentSlot -= handler.slots
        }
        return false
    }

    override fun canExtractItem(slot: Int): Boolean {
        var currentSlot = slot
        if (currentSlot < 0) {
            return false
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                if (handler is IFilteredItemHandler) {
                    // filter the inputs
                    return handler.canExtractItem(currentSlot)
                }
                return true
            }
            currentSlot -= handler.slots
        }
        return false

    }

    override val inventories: Int
        get() = this.handlers.size

    override fun getInventory(inventory: Int): IItemHandler {
        return this.handlers[inventory]
    }

    override fun getFilteredInventory(inventory: Int): IFilteredItemHandler? {
        val handler = this.getInventory(inventory)
        if (handler is IFilteredItemHandler) {
            return handler
        }
        return null
    }

    override fun canInsertItem(inventory: Int, slot: Int, stack: ItemStack): Boolean {
        val handler = this.getFilteredInventory(inventory)
        if (handler != null) {
            return handler.canInsertItem(slot, stack)
        }
        return true
    }

    override fun canExtractItem(inventory: Int, slot: Int): Boolean {
        val handler = this.getFilteredInventory(inventory)
        if (handler != null) {
            return handler.canExtractItem(slot)
        }
        return true
    }

    override fun getSlots() =
        this.handlers.indices
            .map { this.handlers[it] }
            .sumBy { it.slots }

    override fun getStackInSlot(slot: Int): ItemStack {
        var currentSlot = slot
        if (currentSlot < 0) {
            return ItemStack.EMPTY
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                return handler.getStackInSlot(currentSlot)
            }
            currentSlot -= handler.slots
        }
        return ItemStack.EMPTY
    }

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        var currentSlot = slot
        if (currentSlot < 0) {
            return ItemStack.EMPTY
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                return handler.insertItem(currentSlot, stack, simulate)
            }
            currentSlot -= handler.slots
        }
        return ItemStack.EMPTY
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        var currentSlot = slot
        if (currentSlot < 0) {
            return ItemStack.EMPTY
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                return handler.extractItem(currentSlot, amount, simulate)
            }
            currentSlot -= handler.slots
        }
        return ItemStack.EMPTY
    }

    override fun getSlotLimit(slot: Int): Int {
        var currentSlot = slot
        if (currentSlot < 0) {
            return 0
        }
        for (i in this.handlers.indices) {
            val handler = this.handlers[i]
            if (handler.slots > currentSlot) {
                return handler.getSlotLimit(currentSlot)
            }
            currentSlot -= handler.slots
        }
        return 0
    }

    override fun setStackInSlot(slot: Int, stack: ItemStack) {
        var currentSlot = slot
        if (currentSlot >= 0) {
            for (i in this.handlers.indices) {
                val handler = this.handlers[i]
                if (handler.slots > currentSlot) {
                    if (handler is IItemHandlerModifiable) {
                        handler.setStackInSlot(currentSlot, stack)
                    } else {
                        throw RuntimeException("Target inventory is not an IItemHandlerModifiable.")
                    }
                    return
                }
                currentSlot -= handler.slots
            }
        }
    }
}
